SlackAPI/Request.cs

```diff
@@ -141,20 +141,23 @@ public RequestPath(string requestPath, bool isPrimaryAPI = true)
             UsePrimaryAPI = isPrimaryAPI;
         }
 
-        static Dictionary<Type, RequestPath> paths = new Dictionary<Type, RequestPath>();
+        static ConcurrentDictionary<Type, RequestPath> paths = new ConcurrentDictionary<Type, RequestPath>();
 
         public static RequestPath GetRequestPath<K>()
         {
             Type t = typeof(K);
-            if (paths.ContainsKey(t))
-                return paths[t];
+            if (paths.TryGetValue(t, out var path))
+                return path;
 
             TypeInfo info = t.GetTypeInfo();
 
-            RequestPath path = info.GetCustomAttribute<RequestPath>();
+            path = info.GetCustomAttribute<RequestPath>();
             if (path == null) throw new InvalidOperationException(string.Format("No valid request path for {0}", t.Name));
 
-            paths.Add(t, path);
+            // Some other thread may have already placed the path to the dictionary, 
+            // the original one will remain there and current one will be GCed once it
+            // gets out of scope of this request.
+            paths.TryAdd(t, path);
```

Comment:
Can you replace this with `AddOrUpdate`?

I've actually seen some cases where TryAdd will throw exceptions if the key exists (with ConcurrentDictionary)

Reply:
Are you sure? According to https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/Concurrent/ConcurrentDictionary.cs `AddOrUpdate` will call `TryAddInternal` with pretty much the same arguments as `TryAdd` in case the key is not present in the dictionary (which we assert ourselves before adding it). Of course there is some more processing like another check by the `AddOrUpdate` or calling the update method and updating the value.

How about using `paths[t] = path`?

Reply:
IMHO `AddOrUpdate` will add more complexity and lambda allocation.

Reply:
Implementation in CoreFX is pretty much the same...

Reply:
@Inumedia any feedback?

Reply:
Yeah, I'm not sure if `AddOrUpdate` is better. I need to build a proof that `TryAdd` actually does throw an exception so I can report it & track it against this, I guess in the short-term you could wrap the `TryAdd` in an additional try/catch block

Reply:
And just consume the exception? Do you have a logger that I may use to report it? But I assume, it is better that worrying about it in the long run.

Reply:
I added a `System.Diagnostics.Trace.TraceError` which will make the exception visible in AppInsights on Azure (and other tools). It is however not supported on .NET Standard prior 2.0 so I the older ones just use `System.Diagnostics.Debug.Write`.

Reply:
@Inumedia any thoughts?

Reply:
@Inumedia any chance we can finalize this? ;-)